### PR TITLE
Internal: JS API - Partial refactor. [ED-2084]

### PR DESCRIPTION
**Pull request by iNewLegend**
- JS API command base shouldn't know about the editor, separate the editor commands
- The running of commands and handling of the response should be from the commands manager, not from the command itself
- Added a global trace that combined all commands managers traces (for easy debugging)
- Each command class should know its own component and full command name on a static level.

Next:
- Todo: Route Hooks
- Todo: Async Hooks
- Todo: Maybe call `initialize` after `validateArgs`.
- Todo: Update Docs.

### CODE REVIEW

File: assets/dev/js/editor/document/base/command-history.js [hunk 1]
```diff
@@ -1,13 +1,11 @@
...
-export default class CommandHistory extends CommandBase {
+export default class CommandHistory extends CommandEditor {
 	static getInstanceType() {
 		return 'CommandHistory';
 	}
 
-	constructor( args ) {
-		super( args );
-
+	initialize( args = {} ) {
```
Comment (matipojo) [inline assets/dev/js/editor/document/base/command-history.js:8]: BC, what if someone overrides the constructor?

File: assets/dev/js/editor/document/base/command-disable-enable.js [hunk 2]
```diff
@@ -39,18 +43,16 @@ export default class CommandDisableEnable extends CommandHistory {
...
 		 * @type {string}
 		 */
-		this.type = this.currentCommand === this.constructor.getEnableCommand() ?
+		this.type = this.command === this.constructor.getEnableCommand() ?
 			'enable' : 'disable';
-	}
 
-	getTitle() {
-		elementorModules.ForceMethodImplementation();
+		// Override default logic, since getHistory() depends on `this.type`.
```
Comment (matipojo) [inline assets/dev/js/editor/document/base/command-disable-enable.js:54]: Explain the file

File: assets/dev/js/editor/document/component.js [hunk 3]
```diff
@@ -10,19 +10,19 @@ export default class Component extends ComponentBase {
...
 
 				return this.CommandHistory;
 			},
 			get CommandHistoryDebounceBase() {
-				elementorCommon.helpers.hardDeprecated(
+				elementorCommon.helpers.softDeprecated(
 					'$e.modules.document.CommandHistoryDebounceBase',
-					'3.0.0',
-					'$e.modules.document.CommandHistoryDebounce'
+					'3.0.3',
```
Comment (matipojo) [inline assets/dev/js/editor/document/component.js:24]: Change all deprecations to 3.4.0

File: assets/dev/js/editor/elements/models/element.js [hunk 4]
```diff
@@ -162,7 +162,7 @@ ElementModel = Backbone.Model.extend( {
 	},
 
 	renderRemoteServer: function() {
-		if ( ! this.remoteRender ) {
+		if ( ! elementorCommonConfig.isTesting || ! this.remoteRender ) {
```
Comment (matipojo) [inline assets/dev/js/editor/elements/models/element.js:165]: Moke the method in the test

File: assets/dev/js/editor/regions/panel/commands/change-device-mode.js [hunk 5]
```diff
@@ -1,6 +1,6 @@
-import CommandBase from 'elementor-api/modules/command-base';
+import Command from 'elementor-api/modules/command';
```
Comment (matipojo) [inline assets/dev/js/editor/regions/panel/commands/change-device-mode.js:1]: Replace back to command-base. and rename the command to command-core

File: assets/dev/js/editor/regions/panel/commands/internal/open-default.js [hunk 6]
```diff
@@ -1,6 +1,6 @@
-import CommandInternalBase from 'elementor-api/modules/command-internal-base';
+import CommandInternal from 'elementor-api/modules/command-internal';
```
Comment (matipojo) [inline assets/dev/js/editor/regions/panel/commands/internal/open-default.js:1]: Return the base back

File: assets/dev/js/editor/utils/helpers.js [hunk 7]
```diff
@@ -199,7 +199,7 @@ module.exports = {
 	fetchFa4ToFa5Mapping() {
 		const storageKey = 'fa4Tofa5Mapping';
 		let mapping = elementorCommon.storage.get( storageKey );
-		if ( ! mapping ) {
+		if ( ! elementorCommonConfig.isTesting && ! mapping ) {
```
Comment (matipojo) [inline assets/dev/js/editor/utils/helpers.js:202]: Mock from the test

File: assets/dev/js/modules/imports/args-object.js [hunk 8]
```diff
@@ -16,6 +16,10 @@ export default class ArgsObject extends InstanceType {
 		super();
 
 		this.args = args;
+
+		if ( ! this.instanceTypes ) {
```
Comment (matipojo) [inline assets/dev/js/modules/imports/args-object.js:20]: Explain

File: core/common/assets/js/api/core/commands.js [hunk 9]
```diff
@@ -269,38 +283,139 @@ export default class Commands extends CommandsBackwardsCompatibility {
...
 	 */
 	run( command, args = {} ) {
-		if ( ! this.beforeRun( command, args ) ) {
+		if ( ! this.validateRun( command, args ) ) {
 			return false;
 		}
 
-		const component = this.getComponent( command ),
-			container = component.getRootContainer();
+		this.beforeRun( command, args );
```
Comment (matipojo) [inline core/common/assets/js/api/core/commands.js:290]: `if ( ! this.beforeRun( command, args ) ) {`

File: core/common/assets/js/api/core/commands.js [hunk 10]
```diff
@@ -269,38 +283,139 @@ export default class Commands extends CommandsBackwardsCompatibility {
...
 
-		this.current[ container ] = command;
-		this.currentArgs[ container ] = args;
+		// Call to new command or callback.
+		let instance = this.commands[ command ];
 
-		this.trigger( 'run:before', component, command, args );
+		if ( instance.getInstanceType ) {
```
Comment (matipojo) [inline core/common/assets/js/api/core/commands.js:295]: Extract to `getInstance` method.

File: core/common/assets/js/api/core/components.js [hunk 11]
```diff
@@ -1,4 +1,6 @@
 export default class extends elementorModules.Module {
+	isRegistering = false;
```
Comment (matipojo) [inline core/common/assets/js/api/core/components.js:2]: Explain

File: core/common/assets/js/api/index.js [hunk 12]
```diff
@@ -38,9 +38,28 @@ export default class API {
...
+				return this.Command;
+			},
+
+			get CommandInternalBase() {
+				elementorCommon.helpers.softDeprecated(
+					'$e.modules.CommandInternalBase',
+					'3.0.3',
+					'$e.modules.CommandInternal'
+				);
+
```
Comment (matipojo) [inline core/common/assets/js/api/index.js:57]: Remove

File: core/common/assets/js/api/modules/command-base.js [hunk 13]
```diff
@@ -1,6 +1,8 @@
 import ArgsObject from 'elementor-assets-js/modules/imports/args-object';
 
 export default class CommandBase extends ArgsObject {
+	static registerArgs = {};
```
Comment (matipojo) [inline core/common/assets/js/api/modules/command-base.js:4]: Rename to `config`? `properties`?

File: core/common/assets/js/api/modules/command-base.js [hunk 14]
```diff
@@ -1,6 +1,8 @@
 import ArgsObject from 'elementor-assets-js/modules/imports/args-object';
```
Comment (matipojo) [inline core/common/assets/js/api/modules/command-base.js:1]: Rename `command-infra`

**Review by matipojo — CHANGES_REQUESTED** [review]

File: core/common/assets/js/api/core/commands.js [hunk 15]
```diff
@@ -325,17 +439,43 @@ export default class Commands extends CommandsBackwardsCompatibility {
...
 		delete this.currentArgs[ container ];
 	}
 
+	validateInstanceScope( instance, command, currentComponent ) {
+		if ( ! ( instance instanceof CommandBase ) ) {
+			this.error( `invalid instance, command: '${ command }' ` );
+		}
+
+		// In case of different scope.
+		if ( currentComponent !== instance.component ) {
```
Comment (iNewLegend) [inline core/common/assets/js/api/core/commands.js:468]: Does this behavior happen only in routes?

File: tests/qunit/tests/assets/dev/js/editor/command-bases/command-editor-base.spec.js [hunk 16]
```diff
@@ -0,0 +1,77 @@
...
+	QUnit.module( 'File: assets/dev/js/editor/base/command-editor-base.js', () => {
+		QUnit.module( 'CommandEditorBase', ( hooks ) => {
+			hooks.beforeEach( () => $e.components.isRegistering = true );
+
+			hooks.afterEach( () => $e.components.isRegistering = false );
+
+			QUnit.test( 'constructor(): without containers', ( assert ) => {
+				assert.throws(
+					() => {
+						const instance = new CommandEditorBase( { } );
```
Comment (iNewLegend) [inline tests/qunit/tests/assets/dev/js/editor/command-bases/command-editor-base.spec.js:18]: Fix

File: core/common/assets/js/api/index.js [hunk 17]
```diff
@@ -40,9 +40,18 @@ export default class API {
...
+			// get CommandInternalBase() {
+			// 	elementorCommon.helpers.softDeprecated(
+			// 		'$e.modules.CommandInternalBase',
+			// 		'3.0.3',
+			// 		'$e.modules.CommandInternal'
+			// 	);
+			//
+			// 	return this.CommandInternal;
+			// },
+
```
Comment (iNewLegend) [inline core/common/assets/js/api/index.js:52]:
```suggestion
			// get CommandInternalBase() {
			// 	elementorCommon.helpers.softDeprecated(
			// 		'$e.modules.CommandInternalBase',
			// 		'3.0.3',
			// 		'$e.modules.CommandInternal'
			// 	);
			//
			// 	return this.CommandInternal;
			// },

```

remove

**Comment by iNewLegend** [comment]
@matipojo, I think it's enough for adding/changes features/behaviors, it should be merged, if the bugs will occur it will be too hard to handle.

File: assets/dev/js/editor/document/base/command-history.js (hunk 1 above)
Comment (iNewLegend) [reply assets/dev/js/editor/document/base/command-history.js]: It should call super anyway.

File: assets/dev/js/editor/document/base/command-disable-enable.js (hunk 2 above)
Comment (iNewLegend) [reply assets/dev/js/editor/document/base/command-disable-enable.js]: Handled.

Comment (iNewLegend) [reply assets/dev/js/editor/document/base/command-disable-enable.js]: Handled

File: assets/dev/js/modules/imports/args-object.js (hunk 8 above)
Comment (iNewLegend) [reply assets/dev/js/modules/imports/args-object.js:20]: Handled.

File: core/common/assets/js/api/modules/command-base.js (hunk 13 above)
Comment (iNewLegend) [reply core/common/assets/js/api/modules/command-base.js]: Handled.

File: core/common/assets/js/api/modules/command-base.js (hunk 14 above)
Comment (iNewLegend) [reply core/common/assets/js/api/modules/command-base.js]: Handled.

File: core/common/assets/js/api/index.js (hunk 12 above)
Comment (iNewLegend) [reply core/common/assets/js/api/index.js]: Handled.

File: assets/dev/js/editor/document/component.js (hunk 3 above)
Comment (iNewLegend) [reply assets/dev/js/editor/document/component.js]: Handled.

File: assets/dev/js/editor/regions/panel/commands/change-device-mode.js (hunk 5 above)
Comment (iNewLegend) [reply assets/dev/js/editor/regions/panel/commands/change-device-mode.js]: Handled.

File: core/common/assets/js/api/core/components.js (hunk 11 above)
Comment (iNewLegend) [reply core/common/assets/js/api/core/components.js]: Handled.

File: core/common/assets/js/api/core/commands.js (hunk 10 above)
Comment (iNewLegend) [reply core/common/assets/js/api/core/commands.js:295]: What do you mean? 
```
const { getInstanceType } = instance;
```
?

File: assets/dev/js/editor/elements/models/element.js (hunk 4 above)
Comment (iNewLegend) [reply assets/dev/js/editor/elements/models/element.js:165]: Handled.

File: assets/dev/js/editor/utils/helpers.js (hunk 7 above)
Comment (iNewLegend) [reply assets/dev/js/editor/utils/helpers.js:202]: Handled.

File: core/common/assets/js/api/core/commands.js (hunk 9 above)
Comment (iNewLegend) [reply core/common/assets/js/api/core/commands.js:290]: validateRun is now used also in routes.

File: assets/dev/js/editor/document/command-bases/command-disable-enable.js [hunk 18]
```diff
@@ -1,7 +1,14 @@
-import CommandHistory from './command-history';
+import CommandHistoryBase from './command-history-base';
 import ElementsSettings from 'elementor-document/elements/commands/settings';
 
-export default class CommandDisableEnable extends CommandHistory {
+/**
+ * The class serve as base for commands who needs to 'enable/disable' behavior.
```
Comment (iNewLegend) [inline assets/dev/js/editor/document/command-bases/command-disable-enable.js:5]:
```suggestion
 * The class serve as base for commands who needs the 'enable/disable' behavior.
```

File: core/common/assets/js/api/core/commands.js (hunk 10 above)
Comment (matipojo) [reply core/common/assets/js/api/core/commands.js:290]: `let instance = this.getInstance()`

File: core/common/assets/js/api/core/commands.js [hunk 19]
```diff
@@ -269,38 +278,166 @@ export default class Commands extends CommandsBackwardsCompatibility {
...
-		this.trigger( 'run:before', component, command, args );
+		if ( instance.getInstanceType ) {
+			instance = new instance( args );
+		}
 
-		if ( args.onBefore ) {
-			args.onBefore.apply( component, [ args ] );
+		const currentComponent = this.getComponent( command );
+
+		// Route?.
```
Comment (matipojo) [inline core/common/assets/js/api/core/commands.js:296]:
```suggestion
		// Is simple callback? (e.g.  a route)
```

File: core/common/assets/js/api/core/commands.js [hunk 20]
```diff
@@ -325,13 +462,54 @@ export default class Commands extends CommandsBackwardsCompatibility {
 	 * Method fired after the command runs.
 	 *
 	 * @param {string} command
+	 * @param {{}} args
+	 * @param {*} results
 	 */
-	afterRun( command ) {
-		const component = this.getComponent( command ),
-			container = component.getRootContainer();
+	afterRun( command, args, results = undefined ) {
```
Comment (matipojo) [inline core/common/assets/js/api/core/commands.js:468]: Try rename (we already have runAfter)

**Review by matipojo — CHANGES_REQUESTED** [review]

File: core/common/assets/js/api/core/commands.js (hunk 20 above)
Comment (iNewLegend) [reply core/common/assets/js/api/core/commands.js]: Try now https://github.com/elementor/elementor/pull/14973/commits/73c379b00ba81f1804d35b4707346869982ebe41

File: tests/qunit/tests/assets/dev/js/editor/document/command-bases/command-history-base.spec.js [hunk 21]
```diff
@@ -0,0 +1,81 @@
...
+
+				instance.historyId = Math.random();
+
+				let tempCommand = '',
+					tempArgs = '';
+
+				$e.commandsInternal.on( 'run:before', ( component, command, args ) => {
+					tempCommand = command;
+					tempArgs = args;
+				} );
```
Comment (iNewLegend) [inline tests/qunit/tests/assets/dev/js/editor/document/command-bases/command-history-base.spec.js:41]: where `off()`?

**Comment by iNewLegend** [comment]
Something wrong with the checks.

File: assets/dev/js/editor/command-bases/command-container-base.js [hunk 22]
```diff
@@ -0,0 +1,36 @@
+import CommandBase from 'elementor-api/modules/command-base';
+
+/**
+ * This class should be used only when its 'really' required.
+ * Not evey command in the editor should use this only when you required to have the logic in it.
+ */
+export default class CommandContainerBase extends CommandBase {
```
Comment (iNewLegend) [inline assets/dev/js/editor/command-bases/command-container-base.js:7]:
```suggestion
export default class CommandContainerBase extends CommandBase {
```

File: core/common/assets/js/api/core/commands.js [hunk 23]
```diff
@@ -269,38 +290,203 @@ export default class Commands extends CommandsBackwardsCompatibility {
...
+
+		this.afterRun( instance.command, instance.args, result );
+	}
+
+	/**
+	 * Function handleGetRunAfterPromiseResult().
+	 *
+	 * Responsible to to await for promise result.
+	 * Called on getRunAfter() after runInstance(), to handle promise result.
+	 * awaits the promise, before going to next.
```
Comment (iNewLegend) [inline core/common/assets/js/api/core/commands.js:415]:
```suggestion
	 * awaits the promise, before going next.
```

File: assets/dev/js/editor/document/component.js (hunk 3 above)
Comment (matipojo) [reply assets/dev/js/editor/document/component.js]: Change all depredations to 3.7.0

File: core/common/assets/js/api/core/commands.js [hunk 24]
```diff
@@ -269,38 +290,203 @@ export default class Commands extends CommandsBackwardsCompatibility {
...
+	 *
+	 * Responsible to run everything that need to run after each command run.
+	 * Called on run() after runInstance(), to manipulate results & apply 'after' hooks.
+	 *
+	 * @param {CommandBase} instance
+	 * @param {*} result
+	 *
+	 * @returns {Promise<*>|*}
+	 */
+	getRunAfter( instance, result ) {
```
Comment (matipojo) [inline core/common/assets/js/api/core/commands.js:366]:
```suggestion
	applyRunAfter( instance, result ) {
```

File: core/common/assets/js/api/core/commands.js [hunk 25]
```diff
@@ -269,38 +290,203 @@ export default class Commands extends CommandsBackwardsCompatibility {
...
+		// TODO: Temp code determine if it's a jQuery deferred object.
+		if ( result && 'object' === typeof result && result.promise && result.then && result.fail ) {
+			const handleJQueryDeferred = ( _result ) => {
+				_result.fail( ( e ) => {
+					this.catchApply( e, instance );
+					this.afterRun( instance.command, instance.args, e );
+				} );
+				_result.done( ( __result ) => {
+					this.handleGetRunAfterSimple( instance, __result );
```
Comment (matipojo) [inline core/common/assets/js/api/core/commands.js:375]:
```suggestion
					this.applyRunAfterSync( instance, __result );
```

File: core/common/assets/js/api/core/commands.js [hunk 26]
```diff
@@ -269,38 +290,203 @@ export default class Commands extends CommandsBackwardsCompatibility {
...
+				} );
+
+				return _result;
+			};
+
+			return handleJQueryDeferred( result );
+		} else if ( result instanceof Promise ) {
+			return this.handleGetRunAfterPromiseResult( instance, result );
```
Comment (matipojo) [inline core/common/assets/js/api/core/commands.js:383]:
```suggestion
			return this.applyRunAfterAsync( instance, result );
```

File: core/common/assets/js/api/core/commands.js [hunk 27]
```diff
@@ -269,38 +290,203 @@ export default class Commands extends CommandsBackwardsCompatibility {
...
+		}
+
+		this.handleGetRunAfterSimple( instance, result );
```
Comment (matipojo) [inline core/common/assets/js/api/core/commands.js:386]:
```suggestion
		this.applyRunAfterSync( instance, result );
```

File: core/common/assets/js/api/core/commands.js [hunk 28]
```diff
@@ -269,38 +290,203 @@ export default class Commands extends CommandsBackwardsCompatibility {
...
+	 * Function handleGetRunAfterPromiseResult().
+	 *
+	 * Responsible to to await for promise result.
+	 * Called on getRunAfter() after runInstance(), to handle promise result.
+	 * awaits the promise, before going next.
+	 *
+	 * @param {CommandBase} instance
+	 * @param {*} result
+	 */
+	handleGetRunAfterPromiseResult( instance, result ) {
```
Comment (matipojo) [inline core/common/assets/js/api/core/commands.js:420]:
```suggestion
	applyRunAfterAsync( instance, result ) {
```

File: core/common/assets/js/api/core/commands.js [hunk 29]
```diff
@@ -269,38 +290,203 @@ export default class Commands extends CommandsBackwardsCompatibility {
...
+		// Override initial result ( promise ) to await onAfter promises, first!.
+		return ( async () => {
+			await result.catch( ( e ) => {
+				this.catchApply( e, instance );
+				this.afterRun( instance.command, instance.args, e );
+			} );
+			await result.then( ( _result ) => this.handleGetRunAfterPromiseResultAsync( instance, _result ) );
```
Comment (matipojo) [inline core/common/assets/js/api/core/commands.js:427]:
```suggestion
			await result.then( ( _result ) => this.applyRunAfterAsyncResult( instance, _result ) );
```

File: core/common/assets/js/api/core/routes.js [hunk 30]
```diff
@@ -65,15 +65,25 @@ export default class Routes extends Commands {
...
+	validateRun( route, args = {} ) {
+		if ( ! super.validateRun( route, args ) ) {
 			return false;
 		}
 
 		if ( this.is( route, args ) && ! args.refresh ) {
 			return false;
 		}
 
+		const component = this.getComponent( route );
```
Comment (matipojo) [inline core/common/assets/js/api/core/routes.js:77]: It's not a validation

File: core/common/assets/js/api/core/routes.js [hunk 31]
```diff
@@ -82,11 +92,12 @@ export default class Routes extends Commands {
...
 
-		if ( ! component.isOpen || args.reOpen ) {
-			component.isOpen = component.open( args );
+		if ( args.onBefore ) {
+			args.onBefore.apply( component, [ args ] );
 		}
 
-		return component.isOpen;
+		this.current[ container ] = route;
+		this.currentArgs[ container ] = args;
```
Comment (matipojo) [inline core/common/assets/js/api/core/routes.js:100]: Use `addCurrentTrace`

File: core/common/assets/js/api/modules/command-callback.js [hunk 32]
```diff
@@ -0,0 +1,22 @@
+import CommandBase from './command-base';
+
+export default class CommandCallback extends CommandBase {
```
Comment (matipojo) [inline core/common/assets/js/api/modules/command-callback.js:3]: Explain

File: tests/qunit/editor/bootstrap.js [hunk 33]
```diff
@@ -44,9 +41,44 @@ export default class EditorBootstrap {
 
 		elementor.start();
 
+		this.bypassRemoteBehaviour();
+
 		elementor.$preview.trigger( 'load' );
 	}
 
+	initializeTestUtils() {
+		// Since JS API catch errors that occurs while running commands, the tests should expect it.
```
Comment (matipojo) [inline tests/qunit/editor/bootstrap.js:50]: Rewrite the description

File: tests/qunit/editor/bootstrap.js [hunk 34]
```diff
@@ -44,9 +41,44 @@ export default class EditorBootstrap {
...
+		$e.tests.commands = {
+			exceptCatchApply: ( callback, count = 1 ) => {
+				for ( let i = 0; i < count; ++i ) {
+					exceptCatchApply.push( ( e, instance ) => callback( e, instance ) );
+				}
+			},
+		};
+
+		$e.commands.catchApply = ( e, instance ) => {
+			catchApplyOrig( e, instance );
```
Comment (matipojo) [inline tests/qunit/editor/bootstrap.js:67]: Explain

File: tests/qunit/mock/elments/button.json [hunk 35]
```diff
@@ -248,15 +248,15 @@
...
 			"type": "popover_toggle",
 			"tab": "style",
 			"section": "section_style",
 			"label": "Typography",
 			"return_value": "custom",
 			"render_type": "ui",
 			"name": "typography_typography",
-			"default": "",
-			"groupPrefix": "typography_"
+			"default": ""
```
Comment (matipojo) [inline tests/qunit/mock/elments/button.json:259]: Remove

File: core/common/assets/js/api/modules/command-infra.js [hunk 36]
```diff
@@ -0,0 +1,168 @@
...
+	 * Create Commands Base.
+	 *
+	 * @param [args={}]
+	 */
+	constructor( args = {} ) {
+		super( args );
+
+		if ( args.__manualConstructorHandling ) {
+			return;
+		}
```
Comment (matipojo) [inline core/common/assets/js/api/modules/command-infra.js:58]: Remove

File: tests/qunit/tests/assets/dev/js/editor/command-bases/command-container-base.spec.js [hunk 37]
```diff
@@ -0,0 +1,63 @@
...
+						instance.requireContainer();
+					},
+					new Error( 'container and containers cannot go together please select one of them.' )
+				);
+			} );
+
+			QUnit.test( 'instanceOf(): validation', ( assert ) => {
+				const validateCommandEditor = ( commandEditor ) => {
+					// Base.
+					assert.equal( commandEditor instanceof CommandInfra, true );
```
Comment (matipojo) [inline tests/qunit/tests/assets/dev/js/editor/command-bases/command-container-base.spec.js:40]:
```suggestion
					assert.equal( commandContainer instanceof CommandInfra, true );
```

File: tests/qunit/tests/assets/dev/js/editor/command-bases/command-container-internal.spec.js [hunk 38]
```diff
@@ -0,0 +1,37 @@
...
+import CommandContainerBase from 'elementor-editor/command-bases/command-container-base';
+import CommandContainerInternal from 'elementor-editor/command-bases/command-container-internal';
+
+jQuery( () => {
+	QUnit.module( 'File: assets/dev/js/editor/base/command-container-internal.js', () => {
+		QUnit.module( 'CommandEditorInternal', () => {
+			QUnit.test( 'instanceOf(): validation', ( assert ) => {
+				const validateCommandEditorInternal = ( commandEditorInternal ) => {
+					// Base.
+					assert.equal( commandEditorInternal instanceof CommandInfra, true );
```
Comment (matipojo) [inline tests/qunit/tests/assets/dev/js/editor/command-bases/command-container-internal.spec.js:14]:
```suggestion
					assert.equal( commandContainerInternal instanceof CommandInfra, true );
```

File: tests/qunit/tests/assets/dev/js/editor/command-bases/command-container-base.spec.js [hunk 39]
```diff
@@ -0,0 +1,63 @@
...
+					assert.equal( commandEditor instanceof CommandBase, true );
+					assert.equal( commandEditor instanceof CommandInternalBase, false, );
```
Comment (matipojo) [inline tests/qunit/tests/assets/dev/js/editor/command-bases/command-container-base.spec.js:42]: Remove the `false` checks

File: tests/qunit/tests/assets/dev/js/editor/command-bases/command-container-internal.spec.js [hunk 40]
```diff
@@ -0,0 +1,37 @@
...
+					assert.equal( commandEditorInternal instanceof CommandBase, true );
+					assert.equal( commandEditorInternal instanceof CommandInternalBase, false, );
+					assert.equal( commandEditorInternal instanceof CommandData, false, );
```
Comment (matipojo) [inline tests/qunit/tests/assets/dev/js/editor/command-bases/command-container-internal.spec.js:17]: Remove the `false` checks

File: tests/qunit/tests/assets/dev/js/editor/document/command-bases/command-history-base.spec.js [hunk 41]
```diff
@@ -0,0 +1,85 @@
...
+
+				$e.commandsInternal.off( 'run:before', callback );
+			} );
+
+			QUnit.test( 'instanceOf(): validation', ( assert ) => {
+				const validateHistoryCommand = ( historyCommand ) => {
+					// Base.
+					assert.equal( historyCommand instanceof CommandInfra, true );
+					assert.equal( historyCommand instanceof CommandBase, true );
+					assert.equal( historyCommand instanceof CommandInternalBase, false, );
```
Comment (matipojo) [inline tests/qunit/tests/assets/dev/js/editor/document/command-bases/command-history-base.spec.js:58]: Remove the `false` checks

File: tests/qunit/tests/assets/dev/js/editor/document/command-bases/command-history-debounce-base.spec.js [hunk 42]
```diff
@@ -0,0 +1,44 @@
...
+
+jQuery( () => {
+	QUnit.module( 'File: editor/document/commands/command-history-debounce-base.js', () => {
+		QUnit.module( 'CommandHistoryDebounceBase', () => {
+			QUnit.test( 'instanceOf(): validation', ( assert ) => {
+				const validateHistoryDebounceCommand = ( historyDebounceCommand ) => {
+					// Base.
+					assert.equal( historyDebounceCommand instanceof CommandInfra, true );
+					assert.equal( historyDebounceCommand instanceof CommandBase, true );
+					assert.equal( historyDebounceCommand instanceof CommandInternalBase, false, );
```
Comment (matipojo) [inline tests/qunit/tests/assets/dev/js/editor/document/command-bases/command-history-debounce-base.spec.js:18]: Remove the `false` checks

**Review by matipojo — CHANGES_REQUESTED** [review]

**Comment by iNewLegend** [comment]
@matipojo, please update `elementor:feature/js-api-commands-refactor` from `3.6.0`

File: assets/dev/js/editor/editor-document.js [hunk 43]
```diff
@@ -11,4 +13,33 @@ elementorCommon.elements.$window.on( 'elementor:init-components', () => {
...
+			'$e.modules.editor.document.CommandHistoryBase'
+		);
+
+		return $e.modules.editor.document.CommandHistoryBase;
+	},
+
+	get CommandHistoryDebounce() {
+		elementorCommon.helpers.softDeprecated(
+			'$e.modules.CommandHistoryDebounce',
+			'3.5.0',
```
Comment (iNewLegend) [inline assets/dev/js/editor/editor-document.js:38]:
```suggestion
			'3.6.0',
```

**Pull request merged**